docs: add memory shortage troubleshooting guide (#11148)
VictoriaMetrics/VictoriaMetrics: docs/victoriametrics/BestPractices.md

---
## Install Recommendation

It is recommended to run the latest available release of VictoriaMetrics from [this page](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/latest), as it includes all bug fixes and enhancements.

> There is no need to tune VictoriaMetrics, as it uses reasonable defaults for its command-line flags. These flags are automatically adjusted for the available CPU and RAM resources. There is no need for operating system tuning because VictoriaMetrics is optimized for default OS settings. The only option is to increase the limit on the [number of open files in the OS](https://medium.com/@muhammadtriwibowo/set-permanently-ulimit-n-open-files-in-ubuntu-4d61064429a), so VictoriaMetrics could accept more incoming connections and could keep open more data files. VictoriaMetrics is tested and developed to run efficiently on these defaults, which fit the majority of workloads. Change a setting only when the docs explicitly instruct you to, including when and why.

Reviewer: The including when and why here is not very clear, I would go with something like 

```suggestion
There is no need to tune VictoriaMetrics, as it uses reasonable defaults for its command-line flags. These flags are automatically adjusted for the available CPU and RAM resources. There is no need for operating system tuning because VictoriaMetrics is optimized for default OS settings. The only option is to increase the limit on the [number of open files in the OS](https://medium.com/@muhammadtriwibowo/set-permanently-ulimit-n-open-files-in-ubuntu-4d61064429a), so VictoriaMetrics could accept more incoming connections and could keep open more data files. VictoriaMetrics is tested and developed to run efficiently on these defaults, which fit the majority of workloads. Change a setting only when the docs explicitly instruct you to, and only if your situation matches the conditions they describe.
```

## Memory

VictoriaMetrics components detect the available memory at startup as the smaller of the host RAM and the cgroup memory limit.
To keep them stable:

> 1. Do not set [`GOMEMLIMIT` environment variable](https://pkg.go.dev/runtime#hdr-Environment_Variables). Set the container/cgroup memory limit, and VictoriaMetrics automatically

Reviewer: I prefer to avoid negative sentences, because they are harder to think about. but if we are going to use negative sentences I would try to make very explicit the "not", maybe using uppercase and bold, or just bold. This is important specially here where the not is the key part.

Author: I will keep the negative form here. “Do not set GOMEMLIMIT” is a short and direct rule. Bold or uppercase text would make the page harder to read and would not change the meaning. It is consistent with other VictoriaMetrics docs.

> 1. Do not set [`GOMEMLIMIT` environment variable](https://pkg.go.dev/runtime#hdr-Environment_Variables). Set the container/cgroup memory limit, and VictoriaMetrics automatically

Reviewer: In this line, the container/cgroup memory limit, I understand that only applies to containarized applications, for regular application, you normally don't use the cgroup memory limits. So, maybe we can change to something like it.
```suggestion
1. Do not set [`GOMEMLIMIT` environment variable](https://pkg.go.dev/runtime#hdr-Environment_Variables). If you want to limit the memory usage, we recommend you to use containers and set the container/cgroup memory limit, and VictoriaMetrics automatically
```

Author: I agree that “container/cgroup” is unclear, but I will not add a container recommendation. A Linux process can use a cgroup without a container, and VictoriaMetrics reads the cgroup memory limit directly. I will change this to “cgroup memory limit”.

   sizes its memory-aware limits from it. All VictoriaMetrics components have their own [`GOGC` settings](https://pkg.go.dev/runtime#hdr-Environment_Variables), which work optimally in the majority of cases.

1. Do not hand-tune cache sizes with `-storage.cacheSize*` flags; rely on the defaults.
   If a component needs larger caches, move it to a host with more memory.
   See [Cache tuning](https://docs.victoriametrics.com/victoriametrics/single-server-victoriametrics/#cache-tuning).

> 1. Do not autoscale `vmstorage` with the Vertical Pod Autoscaler (VPA) or the Horizontal Pod Autoscaler (HPA).

Reviewer: The formating in this paragraph is not great, I would prefer to have maybe 2 bullet points there VPA, and HPA, now is all a continues block of text that is hard to understand.

Also, the VPA explanation is a bit hard to understand, is mixing the recreation, and the the in-place resizing, and the reasons why those can affect. Maybe we can reestructure it a bit, talking first, for example about the recreation problem and why affects VM, and then talk about in-place resizing and why it affects VM.

HPA is kinda simpler, but also I think we can make it slightly better, I'll try to make a couple of proposals for thos paragrams later.

   VPA: VictoriaMetrics components calculate cache sizes from the memory limit at startup and do not update them afterwards.
   Modes that recreate the pod (`Recreate`, `Auto`) reset the caches and force a cold start,
   causing slow inserts and query latency spikes. In-place resizing is not picked up at runtime,
   so `vmstorage` keeps the budget and `vm_available_memory_bytes` initialized at startup, which also skews the dashboards.
   Set fixed memory requests and limits for `vmstorage` rather than autoscaling.
   HPA: `vmstorage` is stateful. Adding nodes sends new series to them while existing data stays where it is.
   Removing nodes makes the data on them unavailable to queries and can cause data loss without replication.

>    Frequent scaling keeps changing the routing and can degrade cluster performance and availability.

Reviewer: I'm thinking in something like this, but not necesarilly this.

```suggestion
   VPA. Any mode that recreates the pod (Recreate, Auto) wipes the caches and forces a cold start, which slows down inserts and causes query latency spikes. In-place resizing avoids the restart, but it doesn't help either: VictoriaMetrics calculates cache sizes from the memory limit at startup and never re-reads it. After an in-place resize, vmstorage keeps the memory budget and the vm_available_memory_bytes value it computed at startup, so the caches are never resized and the dashboards report the wrong numbers. Set fixed memory requests and limits for vmstorage instead.

   HPA. vmstorage is stateful, so scaling it changes where data lives. Adding nodes routes new series to them, but existing data is not rebalanced and stays where it was written. Removing nodes makes the data they held unavailable to queries, and without replication that data is lost. Frequent scaling keeps reshuffling the routing, degrading cluster performance and availability.
```

Author: I will split the text into VPA and HPA bullets, but I cannot use this text as written. A pod restart does not always remove all caches. VictoriaMetrics saves many caches during a graceful shutdown. Adding storage nodes changes routing for all newly ingested data, not only for new series. Removing a StatefulSet pod also does not delete its PVC by default. I will replace these claims with the verified behavior.

> 1. Leave headroom for the OS page cache and workload spikes -

Reviewer: It would be useful if we can leave here some kind of number, something like "At least a 10%" or "At least XGB", or whatever makes sense for that headroom, because now is a very hard to follow recomendation. Also, if the system is already automatically detecting the memory, we should mention that this only apply if you are tunning something here, right?

```suggestion
1. Leave enough memory headroom for other system requirements like OS page cache and workload spikes -
```

Author: I will add the documented 50% value. I will not say that this rule applies only when settings are changed. Automatic sizing controls VictoriaMetrics cache limits, but the system still needs free RAM for the OS page cache and workload spikes.

---
## Install Recommendation

It is recommended to run the latest available release of VictoriaMetrics from [this page](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/latest), as it includes all bug fixes and enhancements.

There is no need to tune VictoriaMetrics, as it uses reasonable defaults for its command-line flags. These flags are automatically adjusted for the available CPU and RAM resources. There is no need for operating system tuning because VictoriaMetrics is optimized for default OS settings. The only option is to increase the limit on the [number of open files in the OS](https://medium.com/@muhammadtriwibowo/set-permanently-ulimit-n-open-files-in-ubuntu-4d61064429a), so VictoriaMetrics could accept more incoming connections and could keep open more data files. VictoriaMetrics is tested and developed to run efficiently on these defaults, which fit the majority of workloads. Change a setting only when the docs explicitly instruct you to, including when and why.

## Memory

VictoriaMetrics components detect the available memory at startup as the smaller of the host RAM and the cgroup memory limit.
To keep them stable:

> 1. Do not set `GOMEMLIMIT`. Set the container/cgroup memory limit, and VictoriaMetrics automatically

Reviewer 2: ```suggestion
1. Do not set [`GOMEMLIMIT` environment variable](https://pkg.go.dev/runtime#hdr-Environment_Variables). Set the container/cgroup memory limit, and VictoriaMetrics automatically
```

>    sizes its memory-aware limits from it. All VictoriaMetrics components have their own GC settings,

Reviewer 2: ```suggestion
   sizes its memory-aware limits from it. All VictoriaMetrics components have their own [`GOGC` settings](https://pkg.go.dev/runtime#hdr-Environment_Variables),
```

>    which are recommended.

Reviewer 2: ```suggestion
   which work optimally in the majority of cases.
```

1. Do not hand-tune cache sizes with `-storage.cacheSize*` flags; rely on the defaults.
   If a component needs larger caches, move it to a host with more memory.
   See [Cache tuning](https://docs.victoriametrics.com/victoriametrics/single-server-victoriametrics/#cache-tuning).

1. Do not autoscale `vmstorage` with the Vertical Pod Autoscaler (VPA) or the Horizontal Pod Autoscaler (HPA).

>    VPA: cache sizes are derived from the memory limit, read only once at startup.

Reviewer 2: ```suggestion
   VPA: VictoriaMetrics components calculate cache sizes from the memory limit at startup and do not update them aftewards.
```

   Modes that recreate the pod (`Recreate`, `Auto`) reset the caches and force a cold start,
   causing slow inserts and query latency spikes. In-place resizing is not picked up at runtime,
   so `vmstorage` keeps the budget and `vm_available_memory_bytes` initialized at startup, which also skews the dashboards.
   Set fixed memory requests and limits for `vmstorage` rather than autoscaling.
   HPA: `vmstorage` is stateful. Adding nodes sends new series to them while existing data stays where it is.
   Removing nodes makes the data on them unavailable to queries and can cause data loss without replication.

>    Frequent scaling keeps changing the routing and can degrade the cluster.

Reviewer 2: ```suggestion
   Frequent scaling keeps changing the routing and can degrade the cluster performance and availability.
```

## Memory

VictoriaMetrics components detect the available memory at startup as the smaller of the host RAM and the cgroup memory limit.
To keep them stable:

1. Do not set `GOMEMLIMIT`. VictoriaMetrics paces garbage collection with `GOGC` and does not use `GOMEMLIMIT` for sizing.
   `GOMEMLIMIT` bounds only Go runtime memory, not the process's total RSS, off-heap caches, mmap-ed files,
   or the OS page cache, so it is not a reliable way to size VictoriaMetrics containers. It can curb Go heap growth,
   but setting it too low makes the garbage collector run more often than `GOGC` dictates,
   spending in the worst case up to ~50% of CPU time on GC —
   the ceiling enforced by Go's [GC CPU limiter](https://go.dev/doc/gc-guide).

>    Set the container memory limit and VictoriaMetrics sizes its caches and memory-aware limits automatically.

Reviewer 3: > does not use `GOMEMLIMIT` for sizing

It sounds weird. Neither software does, as this setting is for Go runtime and not for software itself. Maybe remove the sentence?

Author: Done, removed that sentence and reworded the GOMEMLIMIT note.

There is no need to tune VictoriaMetrics because it uses reasonable defaults for command-line flags. These flags are automatically adjusted for the available CPU and RAM resources. There is no need in Operating System tuning because VictoriaMetrics is optimized for default OS settings. The only option is to increase the limit on the [number of open files in the OS](https://medium.com/@muhammadtriwibowo/set-permanently-ulimit-n-open-files-in-ubuntu-4d61064429a), so VictoriaMetrics could accept more incoming connections and could keep open more data files.

## Memory

VictoriaMetrics components detect the available memory at startup as the smaller of the host RAM and the cgroup memory limit.
To keep them stable:

1. Do not set `GOMEMLIMIT`. VictoriaMetrics paces garbage collection with `GOGC` and does not use `GOMEMLIMIT` for sizing.
   `GOMEMLIMIT` bounds only Go runtime memory, not the process's total RSS, off-heap caches, mmap-ed files,
   or the OS page cache, so it is not a reliable way to size VictoriaMetrics containers. It can curb Go heap growth,

>    but set too low it makes the garbage collector run more often than `GOGC` dictates,

Reviewer 4: ```suggestion
   but setting too low it makes the garbage collector run more often than `GOGC` dictates,
```

   spending in the worst case up to ~50% of CPU time on GC —
   the ceiling enforced by Go's [GC CPU limiter](https://go.dev/doc/gc-guide).

>    Control memory with the container memory limit plus `-memory.allowedPercent` or `-memory.allowedBytes`.

Reviewer 4: ```suggestion
   Make sure that you keep memory with the container memory limit plus `-memory.allowedPercent` or `-memory.allowedBytes` under control.
```

Reviewer 4: I think we should expand into what "control" means here

Author: I have rephased it

> 1. Do not hand-tune cache sizes with `-storage.cacheSize*` flags — it frequently causes OOM.

Reviewer 5: Why and where does this informations come from?

Author: Hm, I am not sure what your question is about. 
Cache tuning isn't recommended because, when it is needed, we have to determine the cause and may need to adjust it internally in the VictoriaMetrics component rather than patching it on the customer's side. Our main recommendation and global position is that everything should work with a default value for most of our customers. But I can agree that the second part about OOM could be overclaiming. If you're concerned about it, I can rephrase it.

Author: Removed OOM

   If a component needs larger caches, move it to a host with more memory.
   See [Cache tuning](https://docs.victoriametrics.com/victoriametrics/single-server-victoriametrics/#cache-tuning).

> 1. Do not use the Vertical Pod Autoscaler (VPA) for `vmstorage`. Cache sizes are derived from the memory limit read at startup.

Reviewer 4: ```suggestion
1. Do not use the Vertical Pod Autoscaler (VPA) for `vmstorage`. Cache sizes are derived from the memory limit read only once at startup.
```

>    Modes that recreate the pod (`Recreate`, `Auto`) reset the caches and force a cold start,

Reviewer 4: ```suggestion
   Modes that recreate the pod (`Recreate`, `Auto`) make it reset the caches and force a cold start,
```

Author: Fixed

   causing slow inserts and query latency spikes. In-place resizing is not picked up at runtime,

>    so `vmstorage` keeps the budget and `vm_available_memory_bytes` from startup, which also skews the dashboards.

Reviewer 4: ```suggestion
   so `vmstorage` keeps the budget and `vm_available_memory_bytes` initialized at startup, which also skews the dashboards.
```

>    Use fixed memory resources instead.

Reviewer 4: Not sure what that means - avoid changing memory limits as a closing word?

Author: Reworded that part, no longer applies.
